File: src/test/java/org/cip4/xjdf/json/openapi/SchemaSingleton.java
package org.cip4.xjdf.json.openapi;

import com.fasterxml.jackson.core.JsonFactory;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.networknt.schema.*;
import org.cip4.xjdf.json.openapi.model.Schemas;
import org.junit.jupiter.api.Assertions;

import java.io.ByteArrayOutputStream;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.nio.file.Path;
import java.util.Map;

public class SchemaSingleton {

    public static final JsonSchema xjdfSchema;
    public static final JsonSchema xjmfSchema;
    public static final JsonSchema openApiSpec;
    private static final ObjectMapper jsonMapper = new ObjectMapper(new JsonFactory());
    private static final JsonSchemaFactory factory;
    private static final SchemaValidatorsConfig config;

    static {
        try {
            InputStream xjdfStream = OpenApiConverter.class.getResourceAsStream("/xjdf.xsd");
            JsonSchemaConverter jsonConverter = new JsonSchemaConverter(xjdfStream);
            String xjmf, xjdf;
            Schemas xjdfSchemas;
            try (
                ByteArrayOutputStream xjdfOutputStream = new ByteArrayOutputStream();
                ByteArrayOutputStream xjmfOutputStream = new ByteArrayOutputStream()
            ) {
                xjdfSchemas = jsonConverter.convert(xjdfOutputStream, xjmfOutputStream);
                xjmf = xjmfOutputStream.toString(StandardCharsets.UTF_8);
                xjdf = xjdfOutputStream.toString(StandardCharsets.UTF_8);
            }

            OpenApiConverter openApiConverter = new OpenApiConverter();
            String openApi;
            try (ByteArrayOutputStream outputStream = new ByteArrayOutputStream()) {
                openApiConverter.convert(xjdfSchemas, outputStream);
                openApi = outputStream.toString(StandardCharsets.UTF_8);
            }

            SchemaValidatorsConfig.Builder configBuilder = new SchemaValidatorsConfig.Builder();
            config = configBuilder
                .discriminatorKeywordEnabled(true)
                .build();

            Map<String, String> schemas = Map.of(
                JsonSchemaConverter.XJDF_SCHEMA, xjdf,
                JsonSchemaConverter.XJMF_SCHEMA, xjmf,
                OpenApiConverter.SCHEMA, openApi
            );

            factory = JsonSchemaFactory.getInstance(
                SpecVersion.VersionFlag.V202012,
                builder -> builder.schemaLoaders(schemaLoaders -> schemaLoaders.schemas(schemas))
            );

            xjdfSchema = factory.getSchema(SchemaLocation.of(JsonSchemaConverter.XJDF_SCHEMA));
            xjmfSchema = factory.getSchema(SchemaLocation.of(JsonSchemaConverter.XJMF_SCHEMA));
            openApiSpec = factory.getSchema(SchemaLocation.of(OpenApiConverter.SCHEMA), config);
        } catch (Exception e) {
            throw new RuntimeException("Failed to initialize schemas", e);
        }
    }

    private SchemaSingleton() {
    }

    private static JsonNodePath getJsonPathForPath(Path path) {
        String pathString = path.toString().replace("\\", "/");
        if (!pathString.startsWith("/")) {
            pathString = "/" + pathString;
        }

        return new JsonNodePath(PathType.JSON_POINTER)
            .append("paths")
            .append(pathString);
    }

    public static JsonSchema getSchemaForRequest(Path path) {
        return openApiSpec.getRefSchema(
            getJsonPathForPath(path)
                .append("post")
                .append("requestBody")
                .append("content")
                .append("application/vnd.cip4-xjmf+json")
                .append("schema")
        );
    }

    public static JsonSchema getSchemaForResponse(Path path, int code) {
        return openApiSpec.getRefSchema(
            getJsonPathForPath(path)
                .append("post")
                .append("responses")
                .append(Integer.toString(code))
                .append("content")
                .append("application/vnd.cip4-xjmf+json")
                .append("schema")
        );
    }

    public static JsonSchema getSchemaForSignal(Path path) {
        JsonNodePath signalPath = getJsonPathForPath(path)
            .append("post")
            .append("callbacks")
            .append("signal");

        JsonNode signalNode = openApiSpec.getRefSchema(signalPath).getSchemaNode();
        String elementName = signalNode.fieldNames().next();

        return openApiSpec.getRefSchema(
            signalPath
                .append(elementName)
                .append("post")
                .append("requestBody")
                .append("content")
                .append("application/vnd.cip4-xjmf+json")
                .append("schema")
        );
    }

    public static void assertValidJson(JsonSchema schema, String json) {
        try {
            assertValid(schema, jsonMapper.readTree(json));
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Failed to validate JSON", e);
        }

    }

    public static void assertValid(JsonSchema schema, JsonNode json) {
        var result = schema.validate(json);
        Assertions.assertTrue(
            result.isEmpty(),
            json.toPrettyString() + "\n" + String.join("\n", result.stream().map(ValidationMessage::toString).toList())
        );
    }

    public static void assertInvalidJson(JsonSchema schema, String json) {
        try {
            var result = schema.validate(jsonMapper.readTree(json));
            Assertions.assertFalse(result.isEmpty(), json);
        } catch (JsonProcessingException e) {
            throw new RuntimeException("Failed to validate JSON", e);
        }
    }
}

File: src/test/kotlin/org/cip4/xjdf/json/openapi/SchemaSingleton.kt
package org.cip4.xjdf.json.openapi

import com.fasterxml.jackson.databind.JsonNode
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.yaml.YAMLFactory
import com.networknt.schema.*
import org.junit.jupiter.api.Assertions
import java.io.ByteArrayOutputStream
import java.net.URI
import java.nio.file.Path

object SchemaSingleton {

    val xjdfSchema: JsonSchema
    val xjmfSchema: JsonSchema
    val openApiSpec: JsonSchema
    private val typeMap: TypeMap
    private val mapper = ObjectMapper(YAMLFactory())
    private val factory: JsonSchemaFactory
    private val config: SchemaValidatorsConfig

    init {
        val jsonConverter = JsonSchemaConverter(OpenApiConverter::class.java.getResourceAsStream("/xjdf.xsd")!!)
        var xjmf: String
        val xjdf = ByteArrayOutputStream().use { xjdfStream ->
            xjmf = ByteArrayOutputStream().use { xjmfStream ->
                typeMap = jsonConverter.convert(xjdfStream, xjmfStream)
                xjmfStream.toString("UTF-8")
            }
            xjdfStream.toString("UTF-8")
        }

        val openApiConverter = OpenApiConverter(OpenApiConverter::class.java.getResourceAsStream("/xjdf.xsd")!!)
        val openApi = ByteArrayOutputStream().use { outputStream ->
            openApiConverter.convert(typeMap, outputStream)
            outputStream.toString("UTF-8")
        }

        val configbuilder = SchemaValidatorsConfig.Builder()

        config = configbuilder
            .discriminatorKeywordEnabled(true)
            .build()


        val schemas = mapOf(
            JsonSchemaConverter.XJDF_SCHEMA to xjdf,
            JsonSchemaConverter.XJMF_SCHEMA to xjmf,
            OpenApiConverter.SCHEMA to openApi,
        )

        factory = JsonSchemaFactory.getInstance(
            SpecVersion.VersionFlag.V202012
        ) { builder -> builder.schemaLoaders { schemaLoaders -> schemaLoaders.schemas(schemas) } }

        xjdfSchema = factory.getSchema(URI(JsonSchemaConverter.XJDF_SCHEMA))
        xjmfSchema = factory.getSchema(URI(JsonSchemaConverter.XJMF_SCHEMA))
        openApiSpec = factory.getSchema(URI(OpenApiConverter.SCHEMA))
    }

    private fun getSchemaForPath(path: Path): JsonNode {
        var pathString = path.toString().replace("\\", "/")
        if (!pathString.startsWith("/")) {
            pathString = "/$pathString"
        }

        return openApiSpec.schemaNode["paths"][pathString]
    }

    fun getSchemaForRequest(path: Path): JsonSchema {
        val schema = getSchemaForPath(path)["post"]["requestBody"]["content"].first()["schema"]
        return factory.getSchema(schema, config)
    }

    fun getSchemaForResponse(path: Path, code: Int = 200): JsonSchema {
        val schema = getSchemaForPath(path)["post"]["responses"][code.toString()]["content"].first()["schema"]
        return factory.getSchema(schema, config)
    }

    fun getSchemaForSignal(path: Path): JsonSchema {
        val schema =
            getSchemaForPath(path)["post"]["callbacks"]["signal"].first()["post"]["requestBody"]["content"].first()["schema"]
        try {
            return factory.getSchema(schema, config)
        } catch (e: ClassCastException) {
            throw e
        }
    }

    fun assertValid(schema: JsonSchema, json: String) {
        val result = schema.validate(mapper.readTree(json))
        Assertions.assertTrue(
            result.isEmpty(),
            json + "\n" + result.joinToString("\n")
        )
    }

    fun assertInvalid(schema: JsonSchema, json: String) {
        val result = schema.validate(mapper.readTree(json))
        Assertions.assertFalse(result.isEmpty(), json)
    }

}